### ISSUE TITLE
Interpreting the gain values in booster DF

### ISSUE THREAD

**Issue by dthiagarajan** [dmlc/xgboost#9048]
I'm looking at the dataframe returned by
```python3
model._Booster.trees_to_dataframe()
```
for a model that I've set `gamma = 1e-6`, but I see gain values that don't make sense with this gamma value. By my understanding, I would expect that there should be no gain values above -1e-6, but I'm seeing many that are above that value, and even some that are positive. Why might this be the case? Thanks in advance!

**Comment by dthiagarajan** [comment]
Oh hm, I think I was misinterpreting the gain from the dataframe that's given from the Booster dump - I thought those were the changes in the loss function, but looks like it's actually the change in the predictions, i.e. the leaf scores.

Is there some way to get the changes in the loss function per tree?

**Comment by trivialfis** [comment]
it's the change of decision tree split gain. You can find the definition from the original paper (xgboost/CITATION).

**Comment by trivialfis** [comment]
if you are more into source code than equations, here is the loss change: https://github.com/dmlc/xgboost/blob/42d100de188d6f1df29621f8f72c4e429bbfee22/src/tree/hist/evaluate_splits.h#L267 .

**Comment by dthiagarajan** [comment]
Sorry, I think I'm still not quite following, I think there are two different things I want to clarify:
1. The change in the decision tree split gain is the change in the loss, right? If so, after I've trained an XGBRegressor with the sklearn interface, is there a way to access that per tree added?
2. If I have e.g. the dataframe from the booster dump with gain values (-0.025, -0.023, ...), for leaves, those are "scores" on the leaves, correct? Like they're how the prediction would be modified based on the split rules?

**Comment by trivialfis** [comment]
Feel free to reopen if there are further questions.

**Comment by seanv507** [comment]
I would like to reopen?

This is my understanding. 
The 'Cover' column is the sum of Hessians
For leaf nodes, the Gain column is the optimal weight $w^*_j$ equation 5 above. So one can recover the sum of gradients G, by multiplying by corresponding ('Cover' + lambda).
For internal split nodes, the gain column is given by equation 7 in the paper (just below the screenshot pasted above)

Therefore, one should be able to recover the gradients by summing up the gradients of all the leaf nodes associated with an internal split node


I cannot recover the Gain column by doing this.

Here is an example, with $\gamma=0$ and $\lambda=1$
This is a dump from `clf.get_booster.trees_to_dataframe` using the python library.  I added the Hessian, Gradient,Obj  and Expected Gain columns. (formulas in excel sheet)


<google-sheets-html-origin><style type="text/css"></style>
Tree | Node | ID | Feature | Split | Yes | No | Missing | Gain | Cover | Category | Hessian | Gradient | Obj: Gradient^2/(Hessian+lambda) | Expected Gain |   | Lambda | 1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 20 | 0-20 | Leaf |   |   |   |   | 0.335313559 | 2.33931756 |   | 2.33931756 | -1.119718456 | 0.3754567805 |   |   | Gamma | 0
0 | 19 | 0-19 | Leaf |   |   |   |   | 0.0298848841 | 1.40359056 |   | 1.40359056 | -0.07183102531 | 0.002146661866 |   |   |   |  
0 | 15 | 0-15 | f16 | 0.0313699991 | 0-19 | 0-20 | 0-20 | 0.869481564 | 3.74290824 |   | 3.74290812 | -1.191549481 | 0.2993501307 | 0.03912665581 |   |   |  



[Tree.xlsx](https://github.com/user-attachments/files/29840339/Tree.xlsx)

using xgboost scikitlearn interface demo code
```
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split

import xgboost as xgb

X, y = load_breast_cancer(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=94)

# Use "hist" for constructing the trees, with early stopping enabled.
clf = xgb.XGBClassifier(tree_method="hist", early_stopping_rounds=2)
# Fit the model, test sets are used for early stopping.
clf.fit(X_train, y_train, eval_set=[(X_test, y_test)])
```

**Comment by trivialfis** [comment]
Please open a new issue, I can try to trace it down.
